mlkem/poly.c

```diff
@@ -224,34 +224,68 @@ void poly_decompress(poly *r, const uint8_t a[KYBER_POLYCOMPRESSEDBYTES])
+        __CPROVER_loop_invariant(__CPROVER_forall { unsigned int k; (k < i) ==>
+                                 ( r->coeffs[2 * k]     >= 0 && r->coeffs[2 * k]     < KYBER_Q &&
+                                   r->coeffs[2 * k + 1] >= 0 && r->coeffs[2 * k + 1] < KYBER_Q ) })
+        __CPROVER_decreases(KYBER_N / 2 - i)
     {
         // REF-CHANGE: Hoist scalar decompression into separate function
-        r->coeffs[2 * i + 0] = scalar_decompress_q_16((a[0] >> 0) & 0xF);
-        r->coeffs[2 * i + 1] = scalar_decompress_q_16((a[0] >> 4) & 0xF);
-        a += 1;
+        r->coeffs[2 * i + 0] = scalar_decompress_q_16((a[i] >> 0) & 0xF);
+        r->coeffs[2 * i + 1] = scalar_decompress_q_16((a[i] >> 4) & 0xF);
     }
+
+
+
     #elif (KYBER_POLYCOMPRESSEDBYTES == 160)
-    unsigned int j;
-    uint8_t t[8];
     for (i = 0; i < KYBER_N / 8; i++)
+        __CPROVER_assigns(i, __CPROVER_object_whole(r))
+        __CPROVER_loop_invariant(i >= 0)
+        __CPROVER_loop_invariant(i <= KYBER_N / 8)
+        __CPROVER_loop_invariant(__CPROVER_forall { unsigned int k; (k < i) ==>
+                                 ( r->coeffs[8 * k]     >= 0 && r->coeffs[8 * k]     < KYBER_Q &&
+                                   r->coeffs[8 * k + 1] >= 0 && r->coeffs[8 * k + 1] < KYBER_Q &&
+                                   r->coeffs[8 * k + 2] >= 0 && r->coeffs[8 * k + 2] < KYBER_Q &&
+                                   r->coeffs[8 * k + 3] >= 0 && r->coeffs[8 * k + 3] < KYBER_Q &&
+                                   r->coeffs[8 * k + 4] >= 0 && r->coeffs[8 * k + 4] < KYBER_Q &&
+                                   r->coeffs[8 * k + 5] >= 0 && r->coeffs[8 * k + 5] < KYBER_Q &&
+                                   r->coeffs[8 * k + 6] >= 0 && r->coeffs[8 * k + 6] < KYBER_Q &&
+                                   r->coeffs[8 * k + 7] >= 0 && r->coeffs[8 * k + 7] < KYBER_Q ) })
+        __CPROVER_decreases(KYBER_N / 8 - i)
     {
+        uint8_t t;
+        const unsigned int offset = i * 5;
         // REF-CHANGE: Explicitly truncate to avoid warning about
-        // implicit truncation in CBMC.
-        t[0] = 0x1F &  (a[0] >> 0);
-        t[1] = 0x1F & ((a[0] >> 5) | (a[1] << 3));
-        t[2] = 0x1F &  (a[1] >> 2);
-        t[3] = 0x1F & ((a[1] >> 7) | (a[2] << 1));
-        t[4] = 0x1F & ((a[2] >> 4) | (a[3] << 4));
-        t[5] = 0x1F &  (a[3] >> 1);
-        t[6] = 0x1F & ((a[3] >> 6) | (a[4] << 2));
-        t[7] = 0x1F &  (a[4] >> 3);
-        a += 5;
+        // implicit truncation in CBMC and unwind loop for ease
+        // of proof.
 
-        for (j = 0; j < 8; j++)
-        {
-            // REF-CHANGE: Truncation happened before
-            r->coeffs[8 * i + j] = ((uint32_t) t[j] * KYBER_Q + 16) >> 5;
-        }
+        // Decompress 5 8-bit bytes (so 40 bits) into
+        // 8 5-bit values stored in t[]
+        t = 0x1F &  (a[offset + 0] >> 0);
+        r->coeffs[8 * i]     = ((uint32_t) t * KYBER_Q + 16) >> 5;
```

Comment:
What are the obstacles to leaving the code as-is? You say it eases the proof -- but can you get the proof through without making changes?

Reply:
The original code wasn't very pretty to start with.. it assigns 8 elements to a local array, and only then copies that array into r->coeffs in a loop. I will try to restore and prove the original to see how it comes out.

Reply:
Thank you!